Soufianej taint k3s

JOUNAIDSoufiane opened the pull request:
## Added k8s worker taint configuration options
The options to add to site-config are:
- doni_enable_worker_taint: true/false, by default set to false (enables worker tainting through
  the k8s worker on doni devices)
- zun_tolerate_worker_taint: true/false, by default set to false (enables taint enforcement by
  adding tolerations to worker pods)
- k8s_worker_taint
  - key: taint key, by default set to "worker-node"
  - value: taint value, by default set to "true"
  - effect: taint effect, by default set to "NoSchedule"



## Added taint tolerations to core deployments in k3s
k3s defaults.yaml gets the value of k3s_worker_taint from worker_taint
which is defined under kolla/defaults.yml which subsequently defines
defaults for the option and takes in site values from the
k8s_worker_taint option that can be specified through the site config.

Added worker node taint toleration to smarter devices manager
daemonsets.

Furthermore, templated the nvidia device plugin daemonset and added the
toleration there as well.


Taint deployment strategy on a running testbed:

1. Redeploy k3s playbook to apply tolerations to and relaunch the core daemonsets running on the worker nodes
2. Set `zun_tolerate_worker_taint` to `True` and redeploy Zun
3. Finally, set `doni_enable_worker_taint` to `True` and redeploy 

The above sequence ensures that no existing or simultaneous user pods get evicted and inflicts minimal downtime to core daemonsets.

JOUNAIDSoufiane commented:
![](https://github.trello.services/images/mini-trello-icon.png) [When deploying K3S. taint device nodes to only tolerate non-control plane services and add tolerations for device plugins and other device specifics.](https://trello.com/c/wtCFD1Yv/3276-when-deploying-k3s-taint-device-nodes-to-only-tolerate-non-control-plane-services-and-add-tolerations-for-device-plugins-and-oth)

msherman64 commented:
Requires code from:

Zun: https://github.com/ChameleonCloud/zun/pull/20
Doni: https://github.com/ChameleonCloud/doni/pull/138


Order to apply is:
1. pull and deploy new zun
2. pull and deploy new doni
3. deploy this PR

msherman64 commented:
note: add `worker_node_taint` with a `| default(something)` to kolla/defaults.yml

msherman64 commented on roles/k3s/templates/smarter-device-manager-ds-jetson.yaml.j2:
```diff
@@ -69,6 +69,11 @@ spec:
       nodeSelector:
         smarter-device-manager: enabled
         smarter-device-manager/configmap: jetson
+      tolerations:
+        - key: "{{ k3s_worker_taint.split('=')[0] }}"
+        operator: "Equal"
+        value: "{{ k3s_worker_taint.split('=')[1].split(':')[0] }}"
+        effect: "{{ k3s_worker_taint.split(':')[1] }}"
```
It is difficult to maintain jinja templates containing logic.

I suggest putting lines into kolla/defaults.yml instead:
```
k3s_worker_taint_key:
k3s_worker_taint_value:
k3s_worker_taint_effect:
```

If necessary, you can also use those to template the string format, also in defaults.yml
```
# format like 'worker-node=true:NoSchedule'
k3s_worker_taint: "{{ k3s_worker_taint_key }}={{ k3s_worker_taint_value }}:{{ k3s_worker_taint_effect}}"
```

msherman64 commented on roles/k3s/templates/smarter-device-manager-ds-rpi.yaml.j2:
```diff
@@ -67,6 +67,11 @@ spec:
       nodeSelector:
         smarter-device-manager: enabled
         smarter-device-manager/configmap: rpi
+      tolerations:
+        - key: "{{ k3s_worker_taint.split('=')[0] }}"
+        operator: "Equal"
+        value: "{{ k3s_worker_taint.split('=')[1].split(':')[0] }}"
+        effect: "{{ k3s_worker_taint.split(':')[1] }}"
```
suggest refactor as per above comment

msherman64 commented on roles/k3s/templates/nvidia-device-plugin.yaml.j2:
```diff
@@ -0,0 +1,74 @@
+        - key: nvidia.com/gpu
+          operator: Exists
+          effect: NoSchedule
+        - key: "{{ k3s_worker_taint.split('=')[0] }}"
+          operator: "Equal"
+          value: "{{ k3s_worker_taint.split('=')[1].split(':')[0] }}"
+          effect: "{{ k3s_worker_taint.split(':')[1] }}"
```
suggest refactor as per above comment

msherman64 commented on roles/k3s/templates/nvidia-device-plugin.yaml.j2:
```diff
@@ -0,0 +1,74 @@
+      volumes:
+        - name: device-plugin
+          hostPath:
+            path: /var/lib/kubelet/device-plugins
```
I would suggest breaking the "vendoring" of this template into a separate commit (same PR) as adding the taints/tolerations.

Otherwise this appears to be a very large change, and isn't immediately clear which lines are relevant to which intent

msherman64 reviewed (changes requested):
overall LGTM, but with the following caveats:

1. the noted refactoring of how the taint/toleration is specified, to avoid repeating logic throughout the jinja templates
2. related to the integration with doni and zun: I would suggest adding two flags:

`doni_enforce_worker_taints: false`
`zun_enforce_worker_taints: false`

Then, separate from the deployed container versions, the operator can first toggle on the zun support, and then toggle on the doni support, to avoid unexpected breakage.

Finally, is there a migration procedure to ensure currently running/deployed containers are not un-scheduled when this is deployed?
Do we have docs on how to trigger doni to deploy the new annotation? How can an operator check if some otherwise "good" nodes aren't available because they are/aren't tainted correctly?

Commit:
Added k8s worker taint configuration options

The options to add to site-config are:
- doni_enable_worker_taint: true/false, by default set to false (enables worker tainting through
  the k8s worker on doni devices)
- zun_tolerate_worker_taint: true/false, by default set to false (enables taint enforcement by
  adding tolerations to worker pods)
- k8s_worker_taint
  - key: taint key, by default set to "worker-node"
  - value: taint value, by default set to "true"
  - effect: taint effect, by default set to "NoSchedule"

Commit:
Added taint tolerations to core deployments in k3s

k3s defaults.yaml gets the value of k3s_worker_taint from worker_taint
which is defined under kolla/defaults.yml which subsequently defines
defaults for the option and takes in site values from the
k8s_worker_taint option that can be specified through the site config.

Added worker node taint toleration to smarter devices manager
daemonsets.

Furthermore, templated the nvidia device plugin daemonset and added the
toleration there as well.

msherman64 commented on kolla/defaults.yml:
```diff
@@ -92,6 +92,7 @@ enable_cinder: no
 
 # Doni
 enable_doni: yes
+doni_enforce_worker_taints: false
```
this seems unused?

JOUNAIDSoufiane replied:
remnant from a merge :(, I'll change it

msherman64 commented on kolla/defaults.yml:
```diff
@@ -345,6 +346,10 @@ enable_k3s: no
 enable_zun_compute_k8s: "{{ enable_zun | bool and enable_k3s | bool }}"
 blazar_enable_device_plugin_k8s: "{{ enable_k3s | bool }}"
+# K8S worker tainting
+zun_enforce_worker_taint: false
```
should this be `zun_tolerate_worker_taint`? afaik it will not prevent scheduling even if no workers are tainted?

JOUNAIDSoufiane replied:
agreed

msherman64 commented on kolla/defaults.yml:
```diff
@@ -345,6 +346,10 @@ enable_k3s: no
+# K8S worker tainting
+zun_enforce_worker_taint: false
+doni_enable_worker_taint: false
+worker_taint: "{{ k8s_worker_taint.key | default('worker-node') }}={{ k8s_worker_taint.value | default('true') }}:{{ k8s_worker_taint.effect | default('NoSchedule') }}"
```
Is this a string or a dict? you reference `worker_taint.key` in doni.conf?

JOUNAIDSoufiane replied:
this is a dict, looks like my merge was a little destructive

msherman64 reviewed (commented):
Added some comments, mostly just naming/nits for consistency and to avoid confusing future me?
